cli/src/main.rs

```diff
@@ -511,10 +537,222 @@ async fn run_test(test_args: TestArgs) -> anyhow::Result<i32> {
     Ok(exit_code)
 }
 
+async fn run_validate(validate_args: ValidateArgs) -> anyhow::Result<i32> {
+    let ValidateArgs {
+        junit_paths,
+        show_warnings,
+    } = validate_args;
+
+    log::info!(
+        "Starting trunk-analytics-cli {} (git={}) rustc={}",
+        env!("CARGO_PKG_VERSION"),
+        env!("VERGEN_GIT_SHA"),
+        env!("VERGEN_RUSTC_SEMVER")
+    );
+
+    let current_dir = std::env::current_dir()
+        .ok()
+        .and_then(|p| p.to_str().map(String::from))
+        .unwrap_or_default();
+    let (file_sets, file_counter) = build_filesets(&current_dir, &junit_paths, None, &None, None)?;
+
+    if file_counter.get_count() == 0 || file_sets.is_empty() {
+        return Err(anyhow::anyhow!("No JUnit files found to validate."));
+    }
+
+    log::info!("");
+    log::info!(
+        "Validating the following {} files matching the provided globs:",
+        file_counter.get_count()
+    );
+    for file_set in &file_sets {
+        log::info!(
+            "  File set ({:?}): {}",
+            file_set.file_set_type,
+            file_set.glob
+        );
+        for file in &file_set.files {
+            log::info!("    {}", file.original_path_rel);
+        }
+    }
+
+    let mut reports: Vec<WithFilePath<Report>> = Vec::new();
+    let mut parse_errors: Vec<WithFilePath<JunitParseError>> = Vec::new();
+    file_sets.iter().try_for_each(|file_set| {
+        file_set.files.iter().try_for_each(|bundled_file| {
+            let path = std::path::Path::new(&bundled_file.original_path);
+            let file = std::fs::File::open(path)?;
+            let file_buf_reader = BufReader::new(file);
+            let mut junit_parser = JunitParser::new();
+            junit_parser.parse(file_buf_reader).context(format!(
+                "Encountered unrecoverable error while parsing file: {}",
+                bundled_file.original_path_rel
+            ))?;
+            parse_errors.extend(junit_parser.errors().iter().map(|e| WithFilePath::<
+                JunitParseError,
+            > {
+                file_path: bundled_file.original_path_rel.clone(),
+                wrapped: *e,
+            }));
+            reports.extend(junit_parser.into_reports().iter().map(
+                |report| WithFilePath::<Report> {
+                    file_path: bundled_file.original_path_rel.clone(),
+                    wrapped: report.clone(),
+                },
+            ));
+            Ok::<(), anyhow::Error>(())
+        })?;
+        Ok::<(), anyhow::Error>(())
+    })?;
```

Comment:
nit: this is more idiomatically a `fold` operation. `for_each` is typically for side-effects